subiquity/controllers/snaplist.py

```diff
@@ -0,0 +1,236 @@
+
+from subiquity.models.snaplist import SnapSelection
+from subiquity.ui.views.snaplist import SnapListView
+
+log = logging.getLogger('subiquity.controllers.snaplist')
+
+
+class SnapInfoLoaderState:
+    NOT_RUNNING = 0
+    LOADING_LIST = 1
+    LOADING_INFO = 2
+    LOADED = 3
+    STOPPED = -1
+    FAILED = -2
+
+
+class SampleDataSnapInfoLoader:
+
+    def __init__(self, model, snap_data_dir):
+        self.model = model
+        self.state = SnapInfoLoaderState.NOT_RUNNING
+        self.snap_data_dir = snap_data_dir
+
+    def start(self):
+        snap_find_output = os.path.join(self.snap_data_dir, 'find-output.json')
+        with open(snap_find_output) as fp:
+            self.model.load_find_data(json.load(fp))
+        snap_info_glob = os.path.join(self.snap_data_dir, 'info-*.json')
+        for snap_info_file in glob.glob(snap_info_glob):
+            with open(snap_info_file) as fp:
+                self.model.load_info_data(json.load(fp))
+        self.state = SnapInfoLoaderState.LOADED
+
+    def get_snap_list(self, callback):
+        callback(self.model.get_snap_list())
+
+    def get_snap_info(self, snap, callback):
+        callback()
+
+class SnapdSnapInfoLoader:
+
+    def __init__(self, model, run_in_bg, sock):
+        self.state = SnapInfoLoaderState.NOT_RUNNING
+        self.model = model
+        self.run_in_bg = run_in_bg
+        self.url_base = "http+unix://{}/v2/find?".format(quote_plus(sock))
+        self.session = requests_unixsocket.Session()
+        self.pending_info_snaps = []
+        self.ongoing = {} # {snap:[callbacks]}
+
+    def start(self):
+        self.state = SnapInfoLoaderState.LOADING_LIST
+        log.debug("loading list of snaps")
+        def cb(snap_list):
+            if self.state != SnapInfoLoaderState.LOADING_LIST:
+                return
+            self.state = SnapInfoLoaderState.LOADING_INFO
+            self.pending_info_snaps = snap_list
+            log.debug("fetched list of %s snaps", len(self.pending_info_snaps))
+            self._fetch_next_info()
+        self.ongoing[None] = [cb]
+        self.run_in_bg(self._bg_fetch_list, self._fetched_list)
+
+    def stop(self):
+        self.state = SnapInfoLoaderState.STOPPED
+
+    def _bg_fetch_list(self):
+        return self.session.get(self.url_base + 'section=developers', timeout=60)
+
+    def _fetched_list(self, fut):
+        if self.state == SnapInfoLoaderState.STOPPED:
+            return
+        try:
+            response = fut.result()
+            response.raise_for_status()
+        except requests.exceptions.RequestException:
+            log.exception("loading list of snaps failed")
+            self.state = SnapInfoLoaderState.FAILED
+        else:
+            self.model.load_find_data(response.json())
+        for cb in self.ongoing.pop(None):
+            cb(self.model.get_snap_list())
+
+    def get_snap_list(self, callback):
+        if self.state >= SnapInfoLoaderState.LOADING_INFO:
```

Comment:
self.state in [SnapInfoLoaderState.LOADING_INFO, SnapInfoLoaderState.LOADED_INFO]  
This way if someone adds a new state after loaded where it isn't OK to call get_snap_list() it won't break.

Reply:
I replaced the state enum with a bunch of booleans that encode the answers to the questions we actually want to know. Better?